src/realm/dictionary.cpp

```diff
@@ -0,0 +1,252 @@
+
+Dictionary::Dictionary(const ConstObj& obj, ColKey col_key)
+    : m_obj(obj)
+    , m_col_key(col_key)
+{
+    m_clusters = new DictionaryClusterTree(this, col_key.get_type(), m_obj.get_alloc(), m_obj.get_row_ndx());
+    init_from_parent();
+}
+
+Dictionary::~Dictionary()
+{
+    delete m_clusters;
+}
+
+Dictionary& Dictionary::operator=(const Dictionary& other)
+{
+    if (this != &other) {
+        m_obj = other.m_obj;
+        m_col_key = other.m_col_key;
+
+        if (other.m_clusters) {
+            if (!m_clusters)
+                m_clusters =
+                    new DictionaryClusterTree(this, m_col_key.get_type(), m_obj.get_alloc(), m_obj.get_row_ndx());
+            init_from_parent();
+            m_valid = true;
+        }
+        else {
+            delete m_clusters;
+            m_clusters = nullptr;
+            m_valid = false;
+        }
+    }
+    return *this;
+}
+size_t Dictionary::size() const
+{
+    if (!is_attached())
+        return 0;
+    update_if_needed();
+    if (!m_clusters)
+        return 0;
+
+    return m_clusters->size();
+}
+
+Mixed Dictionary::get(Mixed key) const
+{
+    update_if_needed();
+    if (m_valid) {
+        auto hash = key.hash();
+        ObjKey k(int64_t(hash & 0x7FFFFFFFFFFFFFFF));
+        auto s = m_clusters->get(k);
+        ArrayMixed values(m_obj.get_alloc());
+        ref_type ref = to_ref(Array::get(s.mem.get_addr(), 2));
+        values.init_from_ref(ref);
+
+        return values.get(s.index);
+    }
+    throw std::out_of_range("Key not found");
+    return {};
+}
+
+Dictionary::Iterator Dictionary::begin() const
+{
+    return Iterator(this, 0);
+}
+
+Dictionary::Iterator Dictionary::end() const
+{
+    return Iterator(this, size());
+}
+
+void Dictionary::create()
+{
+    if (!m_valid && m_obj.is_valid()) {
+        auto ref = m_clusters->create();
+        m_obj.set_int(m_col_key, from_ref(ref));
+        m_valid = true;
+    }
+}
+
+std::pair<Dictionary::Iterator, bool> Dictionary::insert(Mixed key, Mixed value)
+{
+    REALM_ASSERT(key.get_type() == DataType(m_col_key.get_type()));
+    create();
+    auto hash = key.hash();
+    ObjKey k(int64_t(hash & 0x7FFFFFFFFFFFFFFF));
+    m_obj.bump_content_version();
+    try {
+        auto state = m_clusters->insert(k, key, value);
+        return {Iterator(this, state.index), true};
+    }
+    catch (...) {
+        auto state = m_clusters->get(k);
+        ArrayMixed values(m_obj.get_alloc());
+        ref_type ref = to_ref(Array::get(state.mem.get_addr(), 2));
+        values.init_from_ref(ref);
+        values.set(state.index, value);
+        return {Iterator(this, state.index), false};
+    }
+}
+
+auto Dictionary::operator[](Mixed key) -> MixedRef
+{
+    create();
+    auto hash = key.hash();
+    ObjKey k(int64_t(hash & 0x7FFFFFFFFFFFFFFF));
+    m_obj.bump_content_version();
+    ClusterNode::State state;
+    try {
+        state = m_clusters->insert(k, key, Mixed{});
```

Comment:
there's a bunch of public methods that dereference `m_clusters` without a null check first

Reply:
I don't think so. Dereferencing `m_clusters` is always preceeded by either a call to `create()` or a check on `size()`. Correct me, if I am wrong.